### PR TITLE
[Master] fix 16407 - LS in binding element of object binding pattern

### LINKED ISSUE #16407: Language Service for imported property through dynamic import
Original comment https://github.com/Microsoft/TypeScript/issues/16383#issuecomment-307291358
 
```ts
import('./Foo').then(({ B/*LS*/ar }) => undefined);
```

@ location LS, support language services

- [x] rename

- [x] find all references

- [x]  goto definition

### PR DESCRIPTION
**Pull request by yuit**
Fix #16407 
Because identifier in bindingElement in objectBindingPattern is considered declaration name, previously it will just return its symbol. So when perform goto-definition, it will simply just stay where it is
```ts
function bar<T>(onfulfilled: (value: T) => void) {
  return undefined;
}

interface Test {
  prop2: number
}
bar<Test>(({pr/*gotoDef*/op2})=>{});  // just stay at the same place
```

However, it is desirable to actually goto the `prop2` in an interface. See: https://github.com/Microsoft/TypeScript/compare/master...master-fix16407#diff-3c46922136bdd65e7a2c2d12db87bbe5....

- [x] Updates change in .symbols files and .types file as we will refer the symbol to the object type now

### CODE REVIEW

**Comment by weswigham** [comment]
I've iterated on yui's change - it not longer changes how we match up symbols in the checker, as that was braking flow control analysis. Now, it does for object binding patterns a similar action as it does for property names in an object literal.

File: src/services/goToDefinition.ts [hunk 1]
```diff
@@ -76,6 +76,26 @@ namespace ts.GoToDefinition {
                 declaration => createDefinitionInfo(declaration, shorthandSymbolKind, shorthandSymbolName, shorthandContainerName));
         }
 
+        // If the node is an identifier in bindingelement of ObjectBindingPattern (Note: ArrayBindingPattern can only declaration)
```
Comment (amcasey) [inline src/services/goToDefinition.ts]: Would you mind rephrasing this a little?  I find it a little hard to follow.

File: src/services/goToDefinition.ts [hunk 2]
```diff
@@ -76,6 +76,26 @@ namespace ts.GoToDefinition {
...
+        // instead of just return the declaration symbol which is itself. We should try to get to the original type of the ObjectBindingPattern and return the property declaration.
+        // For example:
+        //      import('./foo').then(({ b/*goto*/ar }) => undefined); => should get use to the declaration in file "./foo"
+        //
+        //      function bar<T>(onfulfilled: (value: T) => void) { //....}
+        //      interface Test {
+        //          pr/*destination*/op1: number
+        //      }
+        //      bar<Test>(({pr/*goto*/op1})=>{});
+        if (isPropertyName(node) && isBindingElement(node.parent) && isObjectBindingPattern(node.parent.parent) && (node.parent.propertyName ? node.parent.propertyName === node : node.parent.name === node)) {
```
Comment (amcasey) [inline src/services/goToDefinition.ts]: Alternatively, `node === (?:)`.

**Comment by weswigham** [comment]
@amcasey done
